functorch/experimental/_map.py

```diff
@@ -45,69 +100,126 @@ def trace_map(proxy_mode, func_overload, f, xs, *args):
 
-@map.py_impl(DispatchKey.Autograd)
-def map_autograd(f, xs, *args):
-    # TODO: support autograd
-    flat_operands, _ = tree_flatten([f, xs, args])
-    assert all(not f.requires_grad for f in flat_operands
-               if isinstance(f, torch.Tensor))
+    with disable_proxy_modes_tracing():
+        xs_slice = _unstack_pytree(mapped_xs)[0]
+        example_args = (*xs_slice, *pos_args)
+        example_xs = example_args[:num_mapped_args]
+        example_pos_args = example_args[num_mapped_args:]
+        example_flat_out, _ = pytree.tree_flatten(f(num_mapped_args, *example_xs, *example_pos_args))
+        example_grad = [torch.ones_like(out) for out in example_flat_out if out is not None and out.requires_grad]
 
-    _ = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.AutogradCPU))
-    return map(f, xs, *args)
+    fw_graph = make_fx(f)(num_mapped_args, *example_xs, *example_pos_args)
 
+    def joint_f(num_mapped, *example_args):
+        joint_mapped_args = example_args[:num_mapped]
+        args = example_args[num_mapped:]
 
-@map.py_impl(ProxyTorchDispatchMode)
-def map_proxy_torch_dispatch_mode(f, xs, *args):
+        mapped_input = joint_mapped_args[:num_mapped_args]
+        mapped_grads = joint_mapped_args[num_mapped_args:]
+
+        def fw_with_masks(num_mapped_args, *args):
+            fw_out = f(num_mapped_args, *args)
+            return fw_out, [True if isinstance(ret, torch.Tensor) and ret.requires_grad else False for ret in fw_out]
+
+        joint = create_joint(fw_with_masks)
+        _, grads = joint([num_mapped_args] + list(mapped_input) + list(args), list(mapped_grads))
+        return grads
+
+    joint_num_mapped = len(example_grad) + len(example_xs)
+    joint_graph = make_fx(joint_f)(joint_num_mapped, *example_xs, *example_grad, *example_pos_args)
+    flat_out = MapAutogradOp.apply(fw_graph, joint_graph, num_mapped_args, *args)
+    return flat_out
+
+
+@map_impl.py_impl(ProxyTorchDispatchMode)
+def map_proxy_torch_dispatch_mode(f, num_mapped, *args):
     mode = _get_current_dispatch_mode()
     assert (mode is not None), "Mode should always be enabled for python fallback key"
     with _pop_mode_temporarily() as mode:
-        res = trace_map(mode, map, f, xs, *args)
-    return res
+        if mode.enable_tracing:
+            return trace_map(mode, map_impl, f, num_mapped, *args)
+        else:
+            return map_impl(f, num_mapped, *args)
 
 
-@map.py_impl(FakeTensorMode)
-def map_fake_tensor_mode(f, xs, *args):
-    outs = [f(x, *args) for x in xs]
-    return outs[0].new_empty([xs.shape[0], *outs[0].shape])
+@map_impl.py_impl(FakeTensorMode)
+def map_fake_tensor_mode(f, num_mapped, *args):
+    xs = args[:num_mapped]
+    pos_args = args[num_mapped:]
+    leading_dims = pytree.tree_map(lambda t: t.shape[0], xs)
+    xs_pytree = _unstack_pytree(xs)
+    example_out = f(num_mapped, *xs_pytree[0], *pos_args)
+    return pytree.tree_map(lambda t: t.expand(leading_dims[0], *t.shape), example_out)
```

Comment:
Maybe silly question: why can't this just be map_dense?

Also, the fake_tensor rule is actually wrong because it makes assumptions about the strides of the output. Making it map_dense would fix the problem

Reply:
yeah, you're right.